guides/source/asset_pipeline.md

```diff
@@ -791,39 +791,44 @@ location ~ ^/assets/ {
 
 ### Local Precompilation
 
-There are several reasons why you might want to precompile your assets locally.
-Among them are:
+Sometimes, you may not want or be able to compile assets on the production
+server. For instance, you may have limited write access to your production
+filesystem, or you may plan to deploy frequently without making any changes to
+your assets.
 
-* You may not have write access to your production file system.
-* You may be deploying to more than one server, and want to avoid
-duplication of work.
-* You may be doing frequent deploys that do not include asset changes.
+In such cases, you can precompile assets _locally_ — that is, add a finalized
+set of compiled, production-ready assets to your source code repository before
+pushing to production. This way, they do not need to be precompiled separately
+on the production server upon each deployment.
 
-Local compilation allows you to commit the compiled files into source control,
-and deploy as normal.
+As above, you can perform this step using
 
-There are three caveats:
+```bash
+$ RAILS_ENV=production rails assets:precompile
+```
 
-* You must not run the Capistrano deployment task that precompiles assets.
-* You must ensure any necessary compressors or minifiers are
-available on your development system.
-* You must change the following application configuration setting:
+Note the following caveats:
 
-In `config/environments/development.rb`, place the following line:
+* If precompiled assets are available, they will be served — even if they no
+  longer match the original (uncompiled) assets, _even on the development
+  server._
 
-```ruby
-config.assets.prefix = "/dev-assets"
-```
+  To ensure that the development server always compiles assets on-the-fly (and
+  thus always reflects the most recent state of the code), the development
+  environment _must be configured to keep precompiled assets in a different
+  location than production does._ Otherwise, any assets precompiled for use in
+  production will clobber requests for them in development (_i.e.,_ subsequent
+  changes you make to assets will not be reflected in the browser).
```

Comment:
I wonder if it might be worth a redundant ", when using local precompilation," before "the development environment" here.

As a slightly more fundamental question, though: is this still true? 😅   Production assets are named with digests, and we no longer write non-digested filenames as well. *Plus* development asset requests now also use digests by default. It seems either one of those should be sufficient to ensure a well-formed development asset request will get the correct file content; if we're going out of our way to use the manifested digests, that seems like something we could change / reconfigure.

Reply:
Thanks for the follow-up!

> is this still true?

I rewrote this section on the basis of personal experimentation with a new Rails app (v5.1.4). My observation was that if a Sprockets manifest is present, and one of the assets listed therein is referenced by a view template (_e.g.,_ via `javascript_include_tag`), then the precompiled version of that asset is the one that will be served, **even if the raw version (_e.g.,_ `application.coffee`) is newer**.

On the other hand, suppose you pass the `debug: true` option to `javascript_include_tag`. Then, rather than referencing a single `application.js` file, the resulting HTML references each separate JavaScript file declared in `application.coffee` in turn. When this happens, each constituent JS file is compiled and served on-the-fly, because no pre-compiled versions of them exist (and thus are not present in the Sprockets manifest).

On the other hand, that appears to contradict what you're saying here:

> Plus development asset requests now also use digests by default.

My experimentation was not exhaustive, so please let me know if I'm mistaken. 

---

> if we're going out of our way to use the manifested digests, that seems like something we could change / reconfigure.

My assumption was that we wanted to reserve the option to serve precompiled assets in development if that's what someone really wanted. I couldn't personally imagine a use case for this, but I'm no expert.

IMO the saner default would be for Sprockets to ignore precompiled assets when both of the following conditions are met:

* we're in the dev environment, and
* the timestamp on the raw asset is newer than that of the precompiled asset.

But I'm just trying to document the existing behavior.